File: app/src/main/java/io/github/fvasco/pinpoi/PlacemarkListActivity.java
package io.github.fvasco.pinpoi;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.graphics.Typeface;
import android.location.Location;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.NavUtils;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.TextView;
import android.widget.Toast;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;

import io.github.fvasco.pinpoi.dao.PlacemarkDao;
import io.github.fvasco.pinpoi.model.PlacemarkSearchResult;
import io.github.fvasco.pinpoi.util.Consumer;
import io.github.fvasco.pinpoi.util.Coordinates;
import io.github.fvasco.pinpoi.util.LocationUtil;
import io.github.fvasco.pinpoi.util.Util;

/**
 * An activity representing a list of Placemarks. This activity
 * has different presentations for handset and tablet-size devices. On
 * handsets, the activity presents a list of items, which when touched,
 * lead to a {@link PlacemarkDetailActivity} representing
 * item details. On tablets, the activity presents the list of items and
 * item details side-by-side using two vertical panes.
 */
public class PlacemarkListActivity extends AppCompatActivity {
    public static final String ARG_LATITUDE = "latitude";
    public static final String ARG_LONGITUDE = "longitude";
    public static final String ARG_RANGE = "range";
    public static final String ARG_FAVOURITE = "favourite";
    public static final String ARG_COLLECTION_IDS = "collectionIds";
    public static final String ARG_NAME_FILTER = "nameFilter";
    public static final String ARG_SHOW_MAP = "showMap";
    // clockwise arrow
    private static final char[] ARROWS = new char[]{
           /*N*/ '\u2191', /*NE*/ '\u2197', /*E*/ '\u2192', /*SE*/ '\u2198',
           /*S*/ '\u2193', /*SW*/ '\u2199', /*W*/ '\u2190', /*NW*/ '\u2196'
    };
    private WebView mapWebView;
    private FloatingActionButton starFab;
    private FloatingActionButton mapFab;
    /**
     * Whether or not the activity is in two-pane mode, i.e. running on a tablet
     * device.
     */
    private boolean mTwoPane;
    private PlacemarkDao placemarkDao;
    private long[] placemarkIdArray;
    private PlacemarkDetailFragment fragment;
    private Coordinates searchCoordinate;
    private int range;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_placemark_list);
        Util.initApplicationContext(getApplicationContext());
        placemarkDao = PlacemarkDao.getInstance().open();
        starFab = (FloatingActionButton) findViewById(R.id.fabStar);
        mapFab = (FloatingActionButton) findViewById(R.id.fabMap);

        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setTitle(getTitle());

        // Show the Up button in the action bar.
        final ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        final View recyclerView = findViewById(R.id.placemark_list);
        mapWebView = (WebView) findViewById(R.id.mapWebView);
        if (savedInstanceState == null
                ? getIntent().getBooleanExtra(ARG_SHOW_MAP, false)
                : savedInstanceState.getBoolean(ARG_SHOW_MAP)) {
            recyclerView.setVisibility(View.GONE);
            setupWebView(mapWebView);
        } else {
            setupRecyclerView((RecyclerView) recyclerView);
            mapWebView.setVisibility(View.GONE);
        }

        if (findViewById(R.id.placemark_detail_container) != null) {
            // The detail container view will be present only in the
            // large-screen layouts (res/values-w900dp).
            // If this view is present, then the
            // activity should be in two-pane mode.
            mTwoPane = true;
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(ARG_SHOW_MAP, mapWebView.getVisibility() == View.VISIBLE);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        placemarkDao.close();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            // This ID represents the Home or Up button. In the case of this
            // activity, the Up button is shown. Use NavUtils to allow users
            // to navigate up one level in the application structure. For
            // more details, see the Navigation pattern on Android Design:
            //
            // http://developer.android.com/design/patterns/navigation.html#up-vs-back
            //
            NavUtils.navigateUpFromSameTask(this);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setupRecyclerView(final @NonNull RecyclerView recyclerView) {
        final RecyclerView.Adapter oldAdapter = recyclerView.getAdapter();
        if (oldAdapter == null || oldAdapter.getItemCount() == 0) {
            final SimpleItemRecyclerViewAdapter adapter = new SimpleItemRecyclerViewAdapter();
            recyclerView.setAdapter(adapter);
            searchPoi(new Consumer<Collection<PlacemarkSearchResult>>() {
                @Override
                public void accept(Collection<PlacemarkSearchResult> placemarks) {
                    // create array in background thread
                    final PlacemarkSearchResult[] placemarksArray =
                            placemarks.toArray(new PlacemarkSearchResult[placemarks.size()]);
                    Util.MAIN_LOOPER_HANDLER.post(new Runnable() {
                        @Override
                        public void run() {
                            adapter.setPlacemarks(placemarksArray);
                        }
                    });
                }
            });
        }
    }

    private void searchPoi(final @NonNull Consumer<Collection<PlacemarkSearchResult>> placemarksConsumer) {
        // load parameters
        final SharedPreferences preferences = getPreferences(MODE_PRIVATE);
        final float latitude = getIntent().getFloatExtra(ARG_LATITUDE, preferences.getFloat(ARG_LATITUDE, Float.NaN));
        final float longitude = getIntent().getFloatExtra(ARG_LONGITUDE, preferences.getFloat(ARG_LONGITUDE, Float.NaN));
        searchCoordinate = new Coordinates(latitude, longitude);
        range = getIntent().getIntExtra(ARG_RANGE, preferences.getInt(ARG_RANGE, 0));
        String nameFilter = getIntent().getStringExtra(ARG_NAME_FILTER);
        if (nameFilter == null) {
            nameFilter = preferences.getString(ARG_NAME_FILTER, null);
        }
        final boolean favourite = getIntent().getBooleanExtra(ARG_FAVOURITE, preferences.getBoolean(ARG_FAVOURITE, false));
        final String nameFilterFinal = nameFilter;

        // read collections id or parse from preference
        long[] collectionIds = getIntent().getLongArrayExtra(ARG_COLLECTION_IDS);
        if (collectionIds == null) {
            final Set<String> stringIds = preferences.getStringSet(ARG_COLLECTION_IDS, Collections.EMPTY_SET);
            collectionIds = new long[stringIds.size()];
            int i = 0;
            for (final String id : stringIds) {
                collectionIds[i] = Long.parseLong(id);
                ++i;
            }
        }

        // create string set of collection id
        // for preference
        // and deferred job
        final Set<String> collectionIdSet = new TreeSet<>();
        final List<Long> collectionIdList = new ArrayList<>(collectionIds.length);
        for (final long id : collectionIds) {
            collectionIdList.add(id);
            collectionIdSet.add(String.valueOf(id));
        }

        // save parameters in preferences
        preferences.edit()
                .putFloat(ARG_LATITUDE, latitude)
                .putFloat(ARG_LONGITUDE, longitude)
                .putInt(ARG_RANGE, range)
                .putBoolean(ARG_FAVOURITE, favourite)
                .putString(ARG_NAME_FILTER, nameFilter)
                .putStringSet(ARG_COLLECTION_IDS, collectionIdSet)
                .apply();

        Util.showProgressDialog(getString(R.string.title_placemark_list), null,
                new Runnable() {
                    @Override
                    public void run() {
                        try {
                            final Collection<PlacemarkSearchResult> placemarks =
                                    placemarkDao.findAllPlacemarkNear(searchCoordinate,
                                            range, nameFilterFinal, favourite, collectionIdList);
                            Log.d(PlacemarkListActivity.class.getSimpleName(), "placemarks.size()=" + placemarks.size());
                            Util.showToast(getString(R.string.n_placemarks_found, placemarks.size()), Toast.LENGTH_LONG);
                            placemarksConsumer.accept(placemarks);

                            // set up placemark id list for left/right swipe in placemark detail
                            placemarkIdArray = new long[placemarks.size()];
                            int i = 0;
                            for (final PlacemarkSearchResult p : placemarks) {
                                placemarkIdArray[i] = p.getId();
                                ++i;
                            }
                        } catch (Exception e) {
                            Log.e(PlacemarkCollectionDetailFragment.class.getSimpleName(), "updatePlacemarkCollection", e);
                            Util.showToast(getString(R.string.error_search, e.getLocalizedMessage()), Toast.LENGTH_LONG);
                        }
                    }
                }, this);
    }

    @JavascriptInterface
    public void openPlacemark(final long placemarkId) {
        if (mTwoPane) {
            Bundle arguments = new Bundle();
            arguments.putLong(PlacemarkDetailFragment.ARG_PLACEMARK_ID, placemarkId);
            arguments.putLongArray(PlacemarkDetailActivity.ARG_PLACEMARK_LIST_ID, placemarkIdArray);
            fragment = new PlacemarkDetailFragment();
            fragment.setArguments(arguments);
            getSupportFragmentManager().beginTransaction()
                    .replace(R.id.placemark_detail_container, fragment)
                    .commit();

            // show fab
            starFab.setVisibility(View.VISIBLE);
            mapFab.setVisibility(View.VISIBLE);
            mapFab.setOnLongClickListener(fragment.longClickListener);
        } else {
            Intent intent = new Intent(this, PlacemarkDetailActivity.class);
            intent.putExtra(PlacemarkDetailFragment.ARG_PLACEMARK_ID, placemarkId);
            intent.putExtra(PlacemarkDetailActivity.ARG_PLACEMARK_LIST_ID, placemarkIdArray);
            this.startActivity(intent);
        }
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        resetStarFabIcon();
    }

    private void resetStarFabIcon() {
        fragment.resetStarFabIcon(starFab);
    }

    public void onStarClick(final View view) {
        fragment.onStarClick(starFab);
    }

    public void onMapClick(final View view) {
        fragment.onMapClick(view);
    }

    private void setupWebView(final WebView mapWebView) {
        mapWebView.addJavascriptInterface(this, "pinpoi");
        final WebSettings webSettings = mapWebView.getSettings();
        webSettings.setJavaScriptEnabled(true);
        webSettings.setCacheMode(WebSettings.LOAD_CACHE_ELSE_NETWORK);
        webSettings.setAllowContentAccess(false);
        webSettings.setAllowFileAccessFromFileURLs(false);
        webSettings.setAllowUniversalAccessFromFileURLs(false);
        webSettings.setGeolocationEnabled(false);
        webSettings.setJavaScriptCanOpenWindowsAutomatically(false);
        webSettings.setSupportMultipleWindows(false);

        searchPoi(new Consumer<Collection<PlacemarkSearchResult>>() {
            @Override
            public void accept(Collection<PlacemarkSearchResult> placemarksSearchResult) {
                final StringBuilder html = new StringBuilder(1024 + placemarksSearchResult.size() * 256);
                final String leafletVersion = "0.7.7";
                int zoom = (int) (Math.log(40_000_000 / range) / Math.log(2));
                if (zoom < 0) zoom = 0;
                else if (zoom > 18) zoom = 18;
                html.append("<html><head><meta charset=\"utf-8\"><meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no\" />")
                        .append("<style>\n" + "body {padding: 0; margin: 0;}\n" + "html, body, #map {height: 100%;}\n" + "</style>")
                        .append("<link rel=\"stylesheet\" href=\"http://cdn.leafletjs.com/leaflet/v" + leafletVersion + "/leaflet.css\" />")
                        .append("<script src=\"http://cdn.leafletjs.com/leaflet/v" + leafletVersion + "/leaflet.js\"></script>")
                        .append("</html>");
                html.append("<body>\n" + "<div id=\"map\"></div>\n" + "<script>");
                html.append("var map = L.map('map').setView([" + searchCoordinate.getLatitude() + "," + searchCoordinate.getLongitude() + "], " + zoom + ");");
                html.append("L.tileLayer('http://{s}.tile.osm.org/{z}/{x}/{y}.png', {" +
                        "attribution: '&copy; <a href=\"http://osm.org/copyright\">OpenStreetMap</a> contributors'" +
                        "}).addTo(map);\n");
                // search center marker
                html.append("L.circle([" + searchCoordinate.getLatitude() + "," + searchCoordinate.getLongitude() + "], 1, {\n" +
                        "color: 'blue',fillOpacity: 1}).addTo(map);\n");
                // search limit circle
                html.append("L.circle([" + searchCoordinate.getLatitude() + "," + searchCoordinate.getLongitude() + "], " + range + "," +
                        " {color: 'red',fillOpacity: 0}).addTo(map);\n");
                for (final PlacemarkSearchResult psr : placemarksSearchResult) {
                    html.append("L.marker([" + psr.getLatitude() + "," + psr.getLongitude() + "]).addTo(map)")
                            .append(".bindPopup(\"")
                            .append("<a href='javascript:pinpoi.openPlacemark(" + psr.getId() + ")'>");
                    if (psr.isFlagged()) html.append("<b>");
                    html.append(Util.escapeJavascript(psr.getName()));
                    if (psr.isFlagged()) html.append("</b>");
                    html.append("</a>")
                            .append("\");\n");
                }
                html.append("</script>\n" + "</body>\n" + "</html>");
                if (BuildConfig.DEBUG)
                    Log.i(PlacemarkListActivity.class.getSimpleName(), "Map HTML " + html);
                final String htmlText = html.toString();
                Util.MAIN_LOOPER_HANDLER.post(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            mapWebView.loadData(htmlText, "text/html", null);
                        } catch (final Throwable e) {
                            Log.e(PlacemarkListActivity.class.getSimpleName(), "mapWebView.loadData", e);
                            Util.showToast(e);
                        }
                    }
                });
            }
        });
    }

    public class SimpleItemRecyclerViewAdapter
            extends RecyclerView.Adapter<SimpleItemRecyclerViewAdapter.ViewHolder> {

        private final DecimalFormat decimalFormat = new DecimalFormat();
        private final StringBuilder stringBuilder = new StringBuilder();
        private final float[] floatArray = new float[2];
        private PlacemarkSearchResult[] placemarks;

        public SimpleItemRecyclerViewAdapter() {
            decimalFormat.setMinimumFractionDigits(1);
            decimalFormat.setMaximumFractionDigits(1);
        }

        public void setPlacemarks(@NonNull final PlacemarkSearchResult[] placemarks) {
            Objects.requireNonNull(placemarks);
            this.placemarks = placemarks;
            notifyDataSetChanged();
        }

        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(android.R.layout.simple_list_item_1, parent, false);
            return new ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(final ViewHolder holder, int position) {
            final PlacemarkSearchResult placemark = placemarks[position];
            holder.placemark = placemark;
            Location.distanceBetween(searchCoordinate.latitude, searchCoordinate.longitude,
                    placemark.getLatitude(), placemark.getLongitude(),
                    floatArray);
            final float distance = floatArray[0];
            // calculate index for arrow
            int arrowIndex = Math.round(floatArray[1] + 45F / 2F);
            if (arrowIndex < 0) {
                arrowIndex += 360;
            }
            arrowIndex = (arrowIndex / 45);
            if (arrowIndex < 0 || arrowIndex >= ARROWS.length) {
                arrowIndex = 0;
            }

            stringBuilder.setLength(0);
            if (distance < 1_000F) {
                stringBuilder.append(Integer.toString((int) distance)).append(" m");
            } else {
                stringBuilder.append(distance < 10_000F
                        ? decimalFormat.format(distance / 1_000F)
                        : Integer.toString((int) distance / 1_000))
                        .append(" ㎞");
            }
            stringBuilder.append(' ').append(ARROWS[arrowIndex])
                    .append("  ").append(placemark.getName());
            holder.view.setText(stringBuilder.toString());
            holder.view.setTypeface(placemark.isFlagged() ? Typeface.DEFAULT_BOLD : Typeface.DEFAULT);

            holder.view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    openPlacemark(holder.placemark.getId());
                }
            });
            holder.view.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View view) {
                    LocationUtil.openExternalMap(holder.placemark, false, view.getContext());
                    return true;
                }
            });
        }

        @Override
        public int getItemCount() {
            return placemarks == null ? 0 : placemarks.length;
        }

        public class ViewHolder extends RecyclerView.ViewHolder {
            public final TextView view;
            public PlacemarkSearchResult placemark;

            public ViewHolder(View view) {
                super(view);
                this.view = (TextView) view.findViewById(android.R.id.text1);
            }

            @Override
            public String toString() {
                return super.toString() + " '" + placemark + "'";
            }
        }
    }

}

File: app/src/main/java/io/github/fvasco/pinpoi/PlacemarkListActivity.kt
package io.github.fvasco.pinpoi

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.graphics.Typeface
import android.location.Location
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.app.NavUtils
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import androidx.recyclerview.widget.RecyclerView
import io.github.fvasco.pinpoi.dao.PlacemarkCollectionDao
import io.github.fvasco.pinpoi.dao.PlacemarkDao
import io.github.fvasco.pinpoi.dao.use
import io.github.fvasco.pinpoi.model.PlacemarkSearchResult
import io.github.fvasco.pinpoi.util.*
import kotlinx.android.synthetic.main.activity_placemark_list.*
import kotlinx.android.synthetic.main.placemark_list.*
import org.osmdroid.tileprovider.tilesource.TileSourceFactory
import org.osmdroid.views.CustomZoomButtonsController
import org.osmdroid.views.MapView
import org.osmdroid.views.overlay.Marker
import org.osmdroid.views.overlay.infowindow.MarkerInfoWindow
import java.text.DecimalFormat
import kotlin.math.abs
import kotlin.math.ln
import kotlin.math.roundToInt


/**
 * An activity representing a list of Placemarks. This activity
 * has different presentations for handset and tablet-size devices. On
 * handsets, the activity presents a list of items, which when touched,
 * lead to a [PlacemarkDetailActivity] representing
 * item details. On tablets, the activity presents the list of items and
 * item details side-by-side using two vertical panes.
 */
class PlacemarkListActivity : AppCompatActivity() {

    private var showMap: Boolean = false
    private var placemarkIdArray: LongArray? = null
    private var fragment: PlacemarkDetailFragment? = null
    private lateinit var searchCoordinate: Coordinates
    private var range: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_placemark_list)
        val preferences = getPreferences(Context.MODE_PRIVATE)
        val latitude =
            intent.getFloatExtra(ARG_LATITUDE, preferences.getFloat(ARG_LATITUDE, Float.NaN))
        val longitude =
            intent.getFloatExtra(ARG_LONGITUDE, preferences.getFloat(ARG_LONGITUDE, Float.NaN))
        searchCoordinate = Coordinates(latitude, longitude)
        range = intent.getIntExtra(ARG_RANGE, preferences.getInt(ARG_RANGE, 0))

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.title = title

        // Show the Up button in the action bar.
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        showMap = savedInstanceState?.getBoolean(
            ARG_SHOW_MAP,
            intent.getBooleanExtra(ARG_SHOW_MAP, preferences.getBoolean(PREFERENCE_SHOW_MAP, false))
        )
            ?: intent.getBooleanExtra(
                ARG_SHOW_MAP,
                preferences.getBoolean(PREFERENCE_SHOW_MAP, false)
            )
        if (showMap) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.INTERNET)
                == PackageManager.PERMISSION_GRANTED
            ) {
                setupWebView(map)
            } else {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(Manifest.permission.INTERNET),
                    PERMISSION_SHOW_MAP
                )
            }
        } else {
            setupRecyclerView(placemarkList as RecyclerView)
        }
        preferences.edit { putBoolean(PREFERENCE_SHOW_MAP, showMap) }
    }

    override fun onResume() {
        super.onResume()
        map.onResume()
    }

    override fun onPause() {
        super.onPause()
        map.onPause()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>, grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSION_SHOW_MAP && grantResults.isNotEmpty()
            && grantResults[0] == PackageManager.PERMISSION_GRANTED
        ) {
            setupWebView(map)
        } else {
            setupRecyclerView(placemarkList as RecyclerView)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(ARG_SHOW_MAP, showMap)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            // This ID represents the Home or Up button. In the case of this
            // activity, the Up button is shown. Use NavUtils to allow users
            // to navigate up one level in the application structure. For
            // more details, see the Navigation pattern on Android Design:
            //
            // https://developer.android.com/design/patterns/navigation.html#up-vs-back
            //
            NavUtils.navigateUpFromSameTask(this)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setupRecyclerView(recyclerView: RecyclerView) {
        recyclerView.visibility = View.VISIBLE
        val oldAdapter = recyclerView.adapter
        if (oldAdapter == null || oldAdapter.itemCount == 0) {
            val adapter = SimpleItemRecyclerViewAdapter()
            recyclerView.adapter = adapter
            searchPoi { placemarks ->
                // create array in background thread
                val placemarksArray = placemarks.toTypedArray()
                runOnUiThread { adapter.setPlacemarks(placemarksArray) }
            }
        }
    }

    private fun setupWebView(map: MapView) {
        val context = map.context
        org.osmdroid.config.Configuration.getInstance().userAgentValue = context.packageName;
        map.visibility = View.VISIBLE
        map.setTileSource(TileSourceFactory.DEFAULT_TILE_SOURCE)
        map.controller.setCenter(searchCoordinate.toGeoPoint())
        map.controller.setZoom((ln((40_000_000.0 / range)) / ln(2.0)).coerceIn(0.0, 18.0))
        map.zoomController.setVisibility(CustomZoomButtonsController.Visibility.SHOW_AND_FADEOUT)
        map.setMultiTouchControls(true)

        searchPoi { placemarksSearchResult ->
            // search center
            map.overlays.add(Marker(map).apply {
                position = searchCoordinate.toGeoPoint()
                setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_CENTER)
                setOnMarkerClickListener { _, _ -> true } // do nothing on click
                icon = resources.getDrawable(R.drawable.map_marker_here, context.theme)
            })

            val collectionNameMap: Map<Long, String> =
                PlacemarkCollectionDao(applicationContext).use { placemarkCollectionDao ->
                    try {
                        placemarkCollectionDao.findAllPlacemarkCollection()
                            .associate { it.id to it.name }
                    } catch (e: Exception) {
                        Log.e(
                            PlacemarkCollectionDetailFragment::class.java.simpleName,
                            "searchPoi progress",
                            e
                        )
                        mapOf()
                    }
                }


            placemarksSearchResult
                .sortedWith(compareBy(PlacemarkSearchResult::flagged).thenBy { !it.note.isNullOrEmpty() })
                .forEach { placemark ->
                    val collectionName = collectionNameMap[placemark.collectionId] ?: ""
                    val marker = Marker(map)
                    marker.id = placemark.id.toString()
                    marker.title = placemark.name
                    marker.snippet = collectionName
                    marker.subDescription = placemark.note
                    marker.position = placemark.coordinates.toGeoPoint()
                    marker.relatedObject = placemark
                    val iconId = when {
                        placemark.flagged -> R.drawable.map_marker_favourite
                        placemark.note != null -> R.drawable.map_marker_note
                        else -> R.drawable.map_marker
                    }
                    val iconResource =
                        resources.getDrawable(iconId, context.theme)
                            .apply {
                                colorFilter = PorterDuffColorFilter(
                                    colorFor(collectionName.hashCode()),
                                    PorterDuff.Mode.SRC_ATOP
                                )
                            }
                    marker.icon = iconResource
                    marker.setAnchor(Marker.ANCHOR_CENTER, Marker.ANCHOR_CENTER)
                    marker.infoWindow = PoiMarker(map)
                    map.overlays.add(marker)
                }
        }
    }

    private fun searchPoi(placemarksConsumer: (Collection<PlacemarkSearchResult>) -> Unit) {
        // load parameters
        val preferences = getPreferences(Context.MODE_PRIVATE)
        val nameFilter: String = intent.getStringExtra(ARG_NAME_FILTER)
            ?: preferences.getString(ARG_NAME_FILTER, null)
            ?: ""
        val favourite =
            intent.getBooleanExtra(ARG_FAVOURITE, preferences.getBoolean(ARG_FAVOURITE, false))

        // read collections id or parse from preference
        val collectionIds = intent.getLongArrayExtra(ARG_COLLECTION_IDS)?.toSet()
            ?: preferences.getStringSet(ARG_COLLECTION_IDS, setOf())?.map(String::toLong)
            ?: emptySet()

        // save parameters in preferences
        preferences.edit {
            putFloat(ARG_LATITUDE, searchCoordinate.latitude)
            putFloat(ARG_LONGITUDE, searchCoordinate.longitude)
            putInt(ARG_RANGE, range).putBoolean(ARG_FAVOURITE, favourite)
            putString(ARG_NAME_FILTER, nameFilter)
            putStringSet(ARG_COLLECTION_IDS, collectionIds.map(Long::toString).toSet())
        }

        showProgressDialog(getString(R.string.title_placemark_list), this) {
            PlacemarkDao(applicationContext).use { placemarkDao ->
                try {
                    val placemarks = placemarkDao.findAllPlacemarkNear(
                        searchCoordinate,
                        range.toDouble(), collectionIds, nameFilter, favourite
                    )
                    Log.d(
                        PlacemarkListActivity::class.java.simpleName,
                        "searchPoi progress placemarks.size()=${placemarks.size}"
                    )
                    runOnUiThread {
                        showToast(
                            getString(R.string.n_placemarks_found, placemarks.size),
                            applicationContext
                        )
                        placemarksConsumer(placemarks)
                    }

                    // set up placemark id list for left/right swipe in placemark detail
                    placemarkIdArray = placemarks.map { it.id }.toLongArray()
                } catch (e: Exception) {
                    Log.e(
                        PlacemarkCollectionDetailFragment::class.java.simpleName,
                        "searchPoi progress",
                        e
                    )
                    runOnUiThread {
                        showLongToast(
                            getString(R.string.error_search, e.message),
                            applicationContext
                        )
                    }
                }
            }
        }
    }

    fun openPlacemark(placemarkId: Long) {
        val intent = Intent(this, PlacemarkDetailActivity::class.java)
        intent.putExtra(PlacemarkDetailFragment.ARG_PLACEMARK_ID, placemarkId)
        intent.putExtra(PlacemarkDetailActivity.ARG_PLACEMARK_LIST_ID, placemarkIdArray)
        this.startActivity(intent)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        resetStarFabIcon()
    }

    private fun resetStarFabIcon() {
        fragment?.resetStarFabIcon(fabStar)
    }

    fun onStarClick(@Suppress("UNUSED_PARAMETER") view: View) {
        fragment?.onStarClick(fabStar)
    }

    fun onMapClick(view: View) {
        fragment?.onMapClick(view)
    }

    inner class SimpleItemRecyclerViewAdapter :
        RecyclerView.Adapter<SimpleItemRecyclerViewAdapter.ViewHolder>() {

        private val decimalFormat = DecimalFormat()
        private val stringBuilder = StringBuilder()
        private val floatArray = FloatArray(2)
        private var placemarks: Array<PlacemarkSearchResult>? = null

        init {
            decimalFormat.minimumFractionDigits = 1
            decimalFormat.maximumFractionDigits = 1
        }

        fun setPlacemarks(placemarks: Array<PlacemarkSearchResult>) {
            this.placemarks = placemarks
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view = LayoutInflater.from(parent.context)
                .inflate(android.R.layout.simple_list_item_1, parent, false)
            return ViewHolder(view)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val placemark = placemarks!![position]
            holder.placemark = placemark
            Location.distanceBetween(
                searchCoordinate.latitude.toDouble(),
                searchCoordinate.longitude.toDouble(),
                placemark.coordinates.latitude.toDouble(),
                placemark.coordinates.longitude.toDouble(),
                floatArray
            )
            val distance = floatArray[0]
            // calculate index for arrow
            var arrowIndex = (floatArray[1] + 45f / 2f).roundToInt()
            if (arrowIndex < 0) {
                arrowIndex += 360
            }
            arrowIndex /= 45
            if (arrowIndex < 0 || arrowIndex >= ARROWS.size) {
                arrowIndex = 0
            }

            stringBuilder.setLength(0)
            stringBuilder.append(position + 1).append(") ")
            if (distance < 1000f) {
                stringBuilder.append(distance.toInt().toString()).append(" m")
            } else {
                stringBuilder.append(
                    if (distance < 10000f)
                        decimalFormat.format((distance / 1000f).toDouble())
                    else
                        (distance.toInt() / 1000).toString()
                ).append(" ㎞")
            }
            stringBuilder.append(' ')
                .append(if (distance <= 1F) CENTER else ARROWS[arrowIndex])
                .append("  ")
                .append(placemark.name)
            holder.view.text = stringBuilder.toString()
            holder.view.typeface = when {
                placemark.flagged && placemark.note != null ->
                    Typeface.defaultFromStyle(Typeface.BOLD_ITALIC)
                placemark.flagged -> Typeface.DEFAULT_BOLD
                placemark.note != null -> Typeface.defaultFromStyle(Typeface.ITALIC)
                else -> Typeface.DEFAULT
            }

            holder.view.setOnClickListener { openPlacemark(holder.placemark!!.id) }
            holder.view.setOnLongClickListener {
                LocationUtil(applicationContext).openExternalMap(holder.placemark!!, false)
                true
            }
        }

        override fun getItemCount() = placemarks?.size ?: 0

        inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
            val view = view.findViewById(android.R.id.text1) as TextView
            var placemark: PlacemarkSearchResult? = null
        }
    }

    private inner class PoiMarker(mapView: MapView) :
        MarkerInfoWindow(R.layout.map_placemark_bublle, mapView) {

        init {
            val btn = mView.findViewById<View>(R.id.bubble_moreinfo) as Button
            btn.background = resources.getDrawable(R.drawable.information, view.context.theme)
            btn.visibility = View.VISIBLE
            btn.setOnClickListener {
                val psr = markerReference.relatedObject as PlacemarkSearchResult
                openPlacemark(psr.id)
            }
        }
    }

    companion object {
        const val ARG_LATITUDE = "latitude"
        const val ARG_LONGITUDE = "longitude"
        const val ARG_RANGE = "range"
        const val ARG_FAVOURITE = "favourite"
        const val ARG_COLLECTION_IDS = "collectionIds"
        const val ARG_NAME_FILTER = "nameFilter"
        const val ARG_SHOW_MAP = "showMap"
        private const val PREFERENCE_SHOW_MAP = "showMap"
        private const val PERMISSION_SHOW_MAP = 1

        // clockwise arrow
        private val ARROWS = charArrayOf(
            '\u2191', // N
            '\u2197', // NE
            '\u2192', // E
            '\u2198', // SE
            '\u2193', // S
            '\u2199', // SW
            '\u2190', // W
            '\u2196' // NW
        )

        // white flag
        private const val CENTER = '\u2690'

        private val markerColors =
            arrayOf(
                0x000080, // navy
                0x0000FF, // blue
                0x008000, // green
                0x008080, // teal
                0x800000, // maroon
                0x800080, // purple
                0x808000, // olive
                0x808080, // gray
                0xA52A2A, // brown
                0xFF0000, // red
                0xFF00FF, // magenta
                0xFFA500, // orange
                0xFFFF00, // yellow
            ).map { it xor 0xFF000000u.toInt() }

        private fun colorFor(value: Int) =
            markerColors[abs(value) % markerColors.size]
    }
}
